### PR TITLE
Reusable outbound webhooks

### PR DESCRIPTION
**Pull request by grosser**
@hdt91 

TODO:
 - deleting a stage leaves the outboundwebhook behind 🤷‍♂

### CODE REVIEW

**Comment by hdt91** [comment]
This PR is the start of the attempt to rework existing outbound webhook feature to support external pre-deployment setup use case. The first task is to move this type of webhook beyond the project scope to make it shareable across different stages in different projects. This should help simplify the setup for many application by reusing common setup hooks.

@zendesk/interface

File: app/controllers/outbound_webhooks_controller.rb [hunk 1]
```diff
@@ -28,6 +44,20 @@ def resources_path
...
   def resource_params
-    super.permit(:stage_id, :url, :username, :password).merge(project: current_project)
+    link = OutboundWebhookStage.new(stage: @stage)
+    super.permit(:url, :username, :password).merge(outbound_webhook_stages: [link])
+  end
+
+  def require_project
+    case action_name
+    when "create"
+      @stage = Stage.find(params[:outbound_webhook].delete(:stage_id))
```
Comment (rbayerl) [inline app/controllers/outbound_webhooks_controller.rb]: This looks backwards

Comment (grosser) [reply app/controllers/outbound_webhooks_controller.rb]: stage_id is a fake attribute ... but yeah we can just submit it as a regular param and avoid this delete magic 👍

**Comment by grosser** [comment]
Expanded the addition UI since a single line won't be enough anymore:
Screen Shot 2019-11-06 at 5 25 14 PM

"Delete" will be an unlink when the hook is global:
Screen Shot 2019-11-06 at 5 25 02 PM

New overview UI that shows all webhooks and how often they are used:
Screen Shot 2019-11-06 at 6 49 00 PM

New menu entry to find that:
Screen Shot 2019-11-06 at 5 05 33 PM

New edit UI for global webhooks (also added auditing):
Screen Shot 2019-11-06 at 5 05 29 PM

File: app/controllers/outbound_webhooks_controller.rb [hunk 2]
```diff
@@ -7,27 +7,55 @@ class OutboundWebhooksController < ResourceController
   before_action :authorize_project_deployer!, except: [:index]
   before_action :set_resource, only: [:show, :edit, :update, :destroy, :new, :create]
 
-  def index
```
Comment (grosser) [inline app/controllers/outbound_webhooks_controller.rb]: index json reply changed from `webhooks` -> `outbound_webhooks` ... also no more nested under `projects/foo/`

File: app/controllers/outbound_webhooks_controller.rb [hunk 3]
```diff
@@ -7,27 +7,55 @@ class OutboundWebhooksController < ResourceController
...
 
   def resources_path
-    [@project, 'webhooks']
+    @project ? [@project, 'webhooks'] : super
   end
 
   def resource_params
-    super.permit(:stage_id, :url, :username, :password).merge(project: current_project)
+    allowed = [:url, :username, :password]
+    allowed << :global if action_name == "create"
```
Comment (grosser) [inline app/controllers/outbound_webhooks_controller.rb]: not allowing to edit global or we could get into weird situations where something is global and then gets made non-global

File: app/controllers/outbound_webhooks_controller.rb [hunk 4]
```diff
@@ -7,27 +7,55 @@ class OutboundWebhooksController < ResourceController
...
+    params = super.permit(*allowed)
+    params[:stages] = [@stage] if @stage && action_name == "create"
```
Comment (grosser) [inline app/controllers/outbound_webhooks_controller.rb]: auto-linking to the stage the request came from

**Comment by rbayerl** [comment]
Would it be possible to add Bearer token auth as an option? That's what we're hoping to use.

**Comment by grosser** [comment]
yes, that's for next PR :)

**Comment by hdt91** [comment]
👍

**Pull request merged**